### DISCUSSION TITLE
Missing metrics

### DISCUSSION THREAD

**Discussion by Isdriai** [deepjavalibrary/djl#1948] [Q&A] [answered]
While I'm training my model, I want to store my model's scores like [here](https://github.com/deepjavalibrary/d2l-java/blob/master/chapter_multilayer-perceptrons/mlp-djl.ipynb)

we can see 

```
trainer.getEvaluators().stream()
                    .forEach(evaluator -> {
                        evaluatorMetrics.put("train_epoch_" + evaluator.getName(), metrics.getMetric("train_epoch_" + evaluator.getName()).stream()
                                            .mapToDouble(x -> x.getValue().doubleValue()).toArray());
                        evaluatorMetrics.put("validate_epoch_" + evaluator.getName(), metrics.getMetric("validate_epoch_" + evaluator.getName()).stream()
                                            .mapToDouble(x -> x.getValue().doubleValue()).toArray());
            });

```

But when I debug my code I see only these entries in metrics:

```
forward 
step
train_all_L2 Loss eth/ip/udp/dns 
backward
train_all_Accuracy 
training-metrics 
```

So my question is how to get somethings like 

```
train_epoch_accuracy
validate_epoch_accuracy
train_epoch_L2 Loss eth/ip/udp/dns
validate_epoch_L2 Loss eth/ip/udp/dns
```

"eth/ip/udp/dns" is not important here, it's just the name I gave to the Loss, I don't think it modifies its behavior

And I make my config object like this:

```    
val config = DefaultTrainingConfig(L2Loss("L2 Loss $pathName"))
            .optOptimizer(sgd) // Optimizer (loss function)
            .optDevices(Engine.getInstance().getDevices(1)) // single GPU
            .addEvaluator(Accuracy())
            .addTrainingListeners(*TrainingListener.Defaults.logging())
```

EDIT:

I use 

``` 
pathTrainer.iterateDataset(multiDataset[pathName, true]).forEachIndexed { index, batch ->
                    println("        entrainement batch $index en cours")
                    EasyTrain.trainBatch(pathTrainer, batch)
                    pathTrainer.step()
                }

```

and 

```
pathTrainer.iterateDataset(multiDataset[pathName, false]).forEachIndexed { index, batch ->
                    println("        validation batch $index en cours")
                    EasyTrain.validateBatch(pathTrainer, batch)
                }

```
for the training and the validation, these two blocks of codes are in the same loop (epochs loop) 

multiDataset is a custom dataset I wrote

EDIT2: 

I checked if 

```
pathTrainer.trainingResult.evaluations
```

had values I wanted but the map was empty

Furthermore, like my dataset, the block in the trainer's model (pathTrainer here) is a custom one

**Comment by zachgk** [comment, marked as answer]
I don't see where you are calling `metrics.addMetric()` which you probably need if you want to add to the metric.

But, you can also add to the metric through the training listener system. It is designed to register certain events in your training and do operations automatically like adding evaluator data to the metrics. You are missing the calls to `onEpoch`, `onTrainingBegin`, and `onTrainingEnd` for the [`TrainingListener`](https://javadoc.io/doc/ai.djl/api/latest/ai/djl/training/listener/TrainingListener.html). You can see an example usage [here](https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/training/EasyTrain.java#L45).

**Reply by Isdriai** [reply]
Indeed, I was doing only 

```

pathTrainer.notifyListeners{listener ->
                    listener.onEpoch(pathTrainer)
                }

```

But with 

```
pathTrainer.notifyListeners{listener ->
                    listener.onEpoch(pathTrainer)
                    listener.onTrainingBegin(pathTrainer)
                    listener.onTrainingEnd(pathTrainer)
                }

```

it works, thanks :)
